scripts/retrieve_databundle_light.py

```diff
@@ -481,39 +483,32 @@ def get_best_bundles_by_category(
     returned_bundles : list
         List of bundles to download
     """
-    # dictionary with the number of match by configuration for tutorial/non-tutorial configurations
-    dict_n_matched = {
-        bname: config_bundles[bname]["n_matched"]
-        for bname in config_bundles
-        if config_bundles[bname]["category"] == category
-        and config_bundles[bname].get("tutorial", False) == tutorial
-        and _check_disabled_by_opt(config_bundles[bname], config_enable) != ["all"]
-    }
+    df_empty = pd.DataFrame(columns=["bundle_name", "bundle_size", "n_matches"])
+    i = 0
+    for bname in config_bundles:
+        if (
+            config_bundles[bname]["category"] == category
+            and config_bundles[bname].get("tutorial", False) == tutorial
+            and _check_disabled_by_opt(config_bundles[bname], config_enable) != ["all"]
+        ):
+            bundles = pd.DataFrame(
+                {"bundle_name": bname, "n_matches": config_bundles[bname]["n_matched"]},
+                index=[i],
+            )
+            i = i + 1
+            df_empty = pd.concat([df_empty, bundles], axis=0)
 
     returned_bundles = []
-
-    # check if non-empty dictionary
-    if dict_n_matched:
-        # if non-empty, then pick bundles until all countries are selected
-        # or no more bundles are found
-        dict_sort = sorted(dict_n_matched.items(), key=lambda d: d[1])
-
-        current_matched_countries = []
-        remaining_countries = set(country_list)
-
-        for d_val in dict_sort:
-            bname = d_val[0]
-
-            cbundle_list = set(config_bundles[bname]["countries"])
-
-            # list of countries in the bundle that are not yet matched
-            intersect = cbundle_list.intersection(remaining_countries)
-
-            if intersect:
-                current_matched_countries.extend(intersect)
-                remaining_countries = remaining_countries.difference(intersect)
-
-                returned_bundles.append(bname)
+    if not df_empty.empty:
+        df_empty = df_empty.sort_values(by="n_matches")
+        for i in df_empty.index:
+            url = list(config_bundles[df_empty.loc[i]["bundle_name"]]["urls"].values())
+            response = requests.head(url[0])
+            bundle_size = response.headers["Content-Length"]
+            df_empty.loc[i, "bundle_size"] = bundle_size
+        df_empty = df_empty[df_empty["n_matches"] == len(country_list)]
+        df_empty = df_empty.sort_values(by=["bundle_size"], ascending=True).head(1)
+        returned_bundles = list(df_empty["bundle_name"])
```

Comment:
This may be slightly revised with the revision above.
I don't get the need for the requests.
The goal would be first, to sort the values as you are doing right but there may be the need to revise the sort by:
```
df_matches["neg_bundle_size"] = -df_matches["bundle_size"]
df_matches.sort_values(by=["n_matches", "neg_bundle_size"], inplace=True, ascending=False)
```

The idea is to have in the first rows the bundles that match most countries but are the smallest.
Using ascending False and the negative value of bundle size should do the trick, may you verify?

I dont get much the logic in the for loop; the logic should be:
- initialize the remaining_countries value to the country list
- then, loop over each row of the sorted dataframe and:
   - compare the intersection between the remaining_countries value and the set `config_bundles[bname]["countries"]` where bname is the value of the current iteration over the index of df_matches.
   - if the intersection is empty, continue the iterations
   - otherwise, add the current bundle to the list of bundles to download and update remaining_countries with the difference between remaining_countries and the intersection.
      The core content of this for loop is already existing in the original code

I understand that not having the CI links working is a challenge...
If you debug locally and check the values in debug mode of the functions the debugging may work anyway though.
To test this function, there is no need to actually download the bundles but verify that the expected data bundles are selected.
Not sure if it is clear, if it is not, please let me know

Reply:
This is a great solution that you sort by n_matches and negative bundle sizes. Do you think it is necessary to loop over country list and check intersection. The filter **df_empty[df_empty["n_matches"] == len(country_list)** is checking whether the bundle contains all of the countries or not.  I can do it with the loop as you wish too.